drivers/nec/display.cr

```diff
@@ -0,0 +1,281 @@
+      send(data, name: "power", delay: 5.seconds)
+    else
+      logger.debug { "-- NEC LCD, requested to power off" }
+      # 4 = Power Off
+      data = MsgType::Command.build(Command::SetPower, 4)
+      send(data, name: "power", delay: 10.seconds, timeout: 10.seconds)
+    end
+  end
+
+  def power?(**options) : Bool
+    data = MsgType::Command.build(Command::PowerQuery)
+    send(data, **options, name: "power?").get
+    self[:power].as_bool
+  end
+
+  def switch_to(input : Input)
+    logger.debug { "-- NEC LCD, requested to switch to: #{input}" }
+    data = MsgType::SetParameter.build(Command::VideoInput, input.value)
+    send(data, name: "input", delay: 6.seconds)
+  end
+
+  enum Audio
+    Audio1      = 1
+    Audio2      = 2
+    Audio3      = 3
+    Hdmi        = 4
+    Tv          = 6
+    DisplayPort = 7
+  end
+
+  def switch_audio(input : Audio)
+    logger.debug { "-- NEC LCD, requested to switch audio to: #{input}" }
+    data = MsgType::SetParameter.build(Command::AudioInput, input.value)
+    send(data, name: "audio")
+  end
+
+  def auto_adjust
+    data = MsgType::SetParameter.build(Command::AutoSetup, 1)
+    send(data, name: "auto_adjust")
+  end
+
+  def brightness(val : Int32)
+    data = MsgType::SetParameter.build(Command::BrightnessStatus, val.clamp(0, 100))
+    send(data, name: "brightness")
+    send(MsgType::Command.build(Command::Save), name: "save", priority: 0)
+  end
+
+  def contrast(val : Int32)
+    data = MsgType::SetParameter.build(Command::ContrastStatus, val.clamp(0, 100))
+    send(data, name: "contrast")
+    send(MsgType::Command.build(Command::Save), name: "save", priority: 0)
+  end
+
+  def volume(val : Int32)
+    data = MsgType::SetParameter.build(Command::VolumeStatus, val.clamp(0, 100))
+    send(data, name: "volume")
+    send(MsgType::Command.build(Command::Save), name: "save", priority: 0)
+  end
+
+  def mute_audio(state : Bool = true, index : Int32 | String = 0)
+    logger.debug { "requested to update mute to #{state}" }
+    data = MsgType::SetParameter.build(Command::MuteStatus, state ? 1 : 0)
+    send(data, name: "mute_audio")
+  end
+
+  def do_poll
+    current_power = power?(priority: 0)
+    logger.debug { "Polling, power = #{current_power}" }
+
+    if current_power
+      mute_status
+      volume_status
+      video_input
+      audio_input
+    end
+  end
+
+  # LCD Response code
+  def received(data, task)
+    task_name = task.try &.name
+    logger.debug { "task is #{task_name}" }
+
+    ascii_string = String.new(data)
+    # Check for valid response
+    if !check_checksum(data)
+      return task.try &.retry("-- NEC LCD, invalid response was: #{ascii_string}")
+    end
+
+    logger.debug { "NEC LCD responded with ascii_string #{ascii_string}" }
+
+    if ascii_string[8..9] == "00"
+      parse_response(ascii_string)
+    elsif ascii_string[10..13] == "00D6" # Annoyingly unique case to deal with power status query
```

Comment:
Rather than splitting on this, take a look at the 5th byte of the header. This is the "message type" which you can parse into the `MsgType` Enum already defined.

Power query does indeed look to need some special handling, but this should make it clearer that's it's due to a command response vs some of other message types that share a common structure. 
![9DBD913D-EF8E-48EE-8139-D8C2724C68EC](https://user-images.githubusercontent.com/843652/104181310-42d55880-545a-11eb-8843-b7440647f86e.jpeg)

Reply:
I've made this commit which hopefully makes things clearer https://github.com/PlaceOS/drivers/pull/50/commits/e86303c89c2085c72a62c07d1cde45142d087397

Reply:
[You are still branching](https://github.com/PlaceOS/drivers/commit/e86303c89c2085c72a62c07d1cde45142d087397#diff-7c4dd3c6e0b44d710bc595343e7b26c65e527df44935f29f97927f6bd22c0904R165) on a couple of bytes that are specific to the power query response. Based on the protocol design there's no guarantee these are unique only to that bit of comms. You'll need to first check the response type (there's a header byte for this), then _if_ its a command reply check the command type and if that's a power query command grab out the value.

Reply:
Ah I see what you mean. I think this should address that

https://github.com/PlaceOS/drivers/pull/50/commits/a890cbad481eb23abb26d67e2bb78b3f373740db